In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [2]:
categorical = pd.read_csv('categorical.csv')
numerical = pd.read_csv('numerical.csv')
target = pd.read_csv('target.csv')

data = pd.concat([categorical,numerical,target],axis=1)

In [3]:
data.head()

STATE  CLUSTER HOMEOWNR GENDER  DATASRCE RFA_2R RFA_2A GEOCODE2 DOMAIN_A  \
0    IL       36        H      F         3      L      E        C        T   
1    CA       14        H      M         3      L      G        A        S   
2    NC       43        U      M         3      L      E        C        R   
3    CA       44        U      F         3      L      E        C        R   
4    FL       16        H      F         3      L      F        A        S   

   DOMAIN_B  ...  MAXRAMNT  LASTGIFT  TIMELAG    AVGGIFT  CONTROLN  HPHONE_D  \
0         2  ...      12.0      10.0        4   7.741935     95515         0   
1         1  ...      25.0      25.0       18  15.666667    148535         0   
2         2  ...      16.0       5.0       12   7.481481     15078         1   
3         2  ...      11.0      10.0        9   6.812500    172556         1   
4         2  ...      15.0      15.0       14   6.864865      7112         1   

   RFA_2F  CLUSTER2  TARGET_B  TARGET_D  
0       4        39         0       0.0  
1       2         1         0       0.0  
2       4        60         0       0.0  
3       4        41         0       0.0  
4       2        26         0       0.0  

[5 rows x 339 columns]

In [7]:
target_d = data[data['TARGET_B']==1]

In [8]:
target_d.shape

(4843, 339)

In [9]:
target_d.head()

STATE  CLUSTER HOMEOWNR GENDER  DATASRCE RFA_2R RFA_2A GEOCODE2 DOMAIN_A  \
20  other       12        H      F         3      L      D        A        S   
30     TX       35        H      M         3      L      D        A        T   
45  other       24        H      F         3      L      D        C        C   
78     CA       13        H      F         2      L      F        A        S   
93     GA       18        H      M         3      L      E        A        S   

    DOMAIN_B  ...  MAXRAMNT  LASTGIFT  TIMELAG    AVGGIFT  CONTROLN  HPHONE_D  \
20         1  ...       7.0       5.0       12   4.066667     82943         1   
30         1  ...      10.0       7.0        9   6.181818    190313         1   
45         1  ...       6.0       5.0        3   4.857143     76585         1   
78         1  ...      17.0      10.0       21  11.000000    156378         0   
93         2  ...      12.0      12.0        6   9.400000     25641         1   

    RFA_2F  CLUSTER2  TARGET_B  TARGET_D  
20       3         3         1       4.0  
30       3        14         1       7.0  
45       3        11         1       5.0  
78       2         2         1      13.0  
93       3        22         1      10.0  

[5 rows x 339 columns]

In [10]:
#X-Y Split, train-test split

X = target_d.drop(['TARGET_B','TARGET_D'],axis=1)
y = target_d['TARGET_D']

In [11]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)

In [24]:
#cat/num X-train
#make a split between the continuous and discrete variables as I don't want to scale or encode the discrete ones, I want to keep their current values
X_train_num = X_train.select_dtypes(include = np.number)
X_train_cat = X_train.select_dtypes(object)

#cat/num X-test
X_test_num = X_test.select_dtypes(include = np.number)
X_test_cat = X_test.select_dtypes(object)


from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import MinMaxScaler
#Scale X-train
transformer = MinMaxScaler().fit(X_train_num)
X_scaled_train = pd.DataFrame(transformer.transform(X_train_num),columns=X_train_num.columns)

#Scale X-test
X_scaled_test = pd.DataFrame(transformer.transform(X_test_num),columns=X_test_num.columns)
X_scaled_test

CLUSTER  DATASRCE  DOMAIN_B  ODATEW_YR  ODATEW_MM    DOB_YR    DOB_MM  \
0    0.000000       0.5  0.000000   0.909091        0.0  0.351648  0.000000   
1    0.865385       0.5  0.333333   0.000000        0.0  0.000000  0.090909   
2    0.846154       1.0  0.333333   0.818182        0.0  0.527473  0.000000   
3    0.538462       1.0  0.333333   0.000000        0.0  0.164835  0.090909   
4    0.846154       1.0  0.333333   0.818182        0.0  0.461538  0.000000   
..        ...       ...       ...        ...        ...       ...       ...   
964  0.673077       1.0  0.333333   0.636364        0.0  0.318681  1.000000   
965  0.653846       1.0  0.000000   0.545455        0.0  0.000000  0.090909   
966  1.000000       1.0  0.666667   0.545455        0.0  0.626374  0.000000   
967  0.250000       1.0  0.000000   0.363636        0.0  0.263736  0.000000   
968  0.346154       1.0  0.333333   0.818182        0.0  0.000000  0.090909   

     MINRDATE_YR  MINRDATE_MM  MAXRDATE_YR  ...  CARDGIFT  MINRAMNT  MAXRAMNT  \
0       0.909091     0.000000     0.909091  ...  0.034483  0.044444  0.015075   
1       0.090909     0.454545     0.818182  ...  0.379310  0.006667  0.010050   
2       0.727273     1.000000     0.818182  ...  0.000000  0.011111  0.015075   
3       0.181818     1.000000     0.545455  ...  0.448276  0.004444  0.001005   
4       0.818182     0.909091     0.818182  ...  0.137931  0.006667  0.001005   
..           ...          ...          ...  ...       ...       ...       ...   
964     0.636364     0.090909     0.727273  ...  0.068966  0.011111  0.010050   
965     0.545455     0.818182     0.818182  ...  0.241379  0.011111  0.010050   
966     0.545455     0.363636     0.909091  ...  0.137931  0.011111  0.020101   
967     0.363636     0.090909     0.818182  ...  0.103448  0.011111  0.010050   
968     0.818182     0.090909     0.818182  ...  0.034483  0.022222  0.011055   

     LASTGIFT   TIMELAG   AVGGIFT  CONTROLN  HPHONE_D    RFA_2F  CLUSTER2  
0    0.044444  0.204545  0.039619  0.908238       0.0  0.000000  0.426230  
1    0.033333  0.113636  0.010920  0.709809       0.0  0.000000  0.885246  
2    0.044444  0.272727  0.022869  0.549203       1.0  0.000000  0.950820  
3    0.011111  0.590909  0.004877  0.231312       0.0  0.666667  0.196721  
4    0.008889  0.113636  0.005001  0.333525       0.0  1.000000  0.704918  
..        ...       ...       ...       ...       ...       ...       ...  
964  0.033333  0.500000  0.021007  0.330735       0.0  0.000000  0.852459  
965  0.022222  0.068182  0.021938  0.683392       0.0  0.333333  0.229508  
966  0.055556  0.250000  0.030686  0.388096       0.0  0.333333  0.950820  
967  0.028889  0.250000  0.017843  0.822848       1.0  0.333333  0.065574  
968  0.035556  0.159091  0.023985  0.806692       0.0  0.000000  0.409836  

[969 rows x 330 columns]

In [25]:
from sklearn.preprocessing import OneHotEncoder
#Create encoder
encoder = OneHotEncoder(drop='first').fit(X_train_cat)

#encode train
encoded_train = encoder.transform(X_train_cat).toarray()
encoded_train = pd.DataFrame(encoded_train)

#encode test
encoded_test = encoder.transform(X_test_cat).toarray()
encoded_test = pd.DataFrame(encoded_test)

In [26]:
train_scaled = pd.concat([encoded_train,X_scaled_train],axis=1)
test_scaled = pd.concat([encoded_test,X_scaled_test],axis=1)

In [28]:
from sklearn.tree import DecisionTreeRegressor
from sklearn.model_selection import cross_val_score
from sklearn.linear_model import LinearRegression
from xgboost import XGBRegressor
from sklearn.ensemble import RandomForestRegressor

DT = DecisionTreeRegressor()
LR = LinearRegression()
xgb = XGBRegressor()
RF = RandomForestRegressor()

model_pipeline = [DT, LR, xgb, RF]
model_names = ['Regression Tree', 'Linear Regression','XGBoost','Random Forest']
scores = {}
i=0
for model in model_pipeline:
    mean_score = np.mean(cross_val_score(model, train_scaled, y_train, cv=10))
    scores[model_names[i]] = mean_score
    i = i+1
print(scores)


{'Regression Tree': -0.20514215492483476, 'Linear Regression': 0.41178956348257645, 'XGBoost': 0.4353661933023787, 'Random Forest': 0.5233272965449235}


In [29]:
val_scores = {}
i=0
for model in model_pipeline:
    model.fit(train_scaled, y_train)
    val_scores[model_names[i]] = model.score(test_scaled,y_test)
    i = i+1
print(val_scores)

{'Regression Tree': -0.26064900702314797, 'Linear Regression': 0.2895529501344033, 'XGBoost': 0.26087543353167375, 'Random Forest': 0.4071674267000196}


In [36]:
from sklearn.metrics import r2_score, mean_squared_error, mean_absolute_error
from math import sqrt

LR =RandomForestRegressor().fit(train_scaled, y_train)
pred = LR.predict(test_scaled)


In [45]:
r2 = r2_score(y_test,pred)
mse = mean_squared_error(y_test,pred)
mae = mean_absolute_error(y_test,pred)
rmse = sqrt(mse)

In [46]:
print('R2 Score:', r2)
print('Mean Squared Error:', mse)
print('Mean aboslute Error:', mae)
print('Root Mean Squared Error:',rmse)

R2 Score: 0.39497301381326866
Mean Squared Error: 115.29677014951496
Mean aboslute Error: 4.874135190918473
Root Mean Squared Error: 10.737633358869866


In [49]:
# how much will the donators give?
round(np.mean(pred),2)

16.36

As above shows we can expect that the average donation amount is 16.36 dollars. 

When you have a look at the calculations I made in the previous lab and update them with predicted donation amount, we get the following:

- Donation amount gained: 183820.96
- Donation amount lost: 114814.48
- Amount spend on marketing: 12449.44
- Amount lost on marketing: 4808.96

Looking at these numbers, we can see that we gain 183.820 dollars with the current model and we soend 12.449. Which means we have a donation gain of 171.371 dollars.
If we wouldn't have made a prediction and instead have send marketing efforts to all of all the 36.228 people (I am using the numbers in the test set), we would have spend 24.635 on marketing efforts and gained 298.635 dollars,leaving a total of 274.000 donations. This actually shows us that it will give us more money to send the marketing efforts to all the possible donators instead of following the model. However, I have not been able to improve the model just yet, so when I have time this weekend I will go back to it and improve the model, so that we lose less donators but also send out less marketing packages. 